### PR TITLE
Adds "files add" http route.

### PR DESCRIPTION
**Pull request by noffle**
Tested to be compatible with a go-ipfs client requesting against the js-ipfs http api.

### CODE REVIEW

**Comment by diasdavid** [comment]
This is pretty rad! Thank you @noffle

File: src/http-api/resources/files.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    fileAdder.on('data', (file) => {
+      serialize.write({
+        Name: file.path,
+        Hash: bs58.encode(file.multihash).toString()
+      })
+      written++
+    })
+
+    fileAdder.on('end', () => {
+      if (written === 0 && file) {
```
Comment (diasdavid) [inline src/http-api/resources/files.js]: is the `file` flag used to signal that at least one file was added?

File: src/http-api/resources/files.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        }).code(500)
+      } else {
+        serialize.end()
+        return reply(serialize)
+          .header('x-chunked-output', '1')
+          .header('content-type', 'application/json')
+      }
+    })
+
+    parser.on('file', (fileName, fileStream) => {
```
Comment (diasdavid) [inline src/http-api/resources/files.js]: isn't fileStream already a readable stream? do we need to copy it?

//cc @xicombd

File: src/http-api/resources/files.js (hunk 1 above)
Comment (nginnever) [inline src/http-api/resources/files.js]: Yeah ```file``` is a boolean that is just used by L77 to generate a no data response from the server. There might be better ways of generating that response if no data is passed to the server in an add call.

Comment (diasdavid) [inline src/http-api/resources/files.js]: a 'no data' response? Like, being `silent`?

Comment (nginnever) [inline src/http-api/resources/files.js]: if the ```file``` flag is true, but the ```written``` var hasn't been incremented, then a 500 server error is returned because the core failed even though the request had valid data parsed.

If at the end of parsing the request, if ```file``` is still false then the request was bad, so ```400 Bad Request``` is returned

Comment (diasdavid) [inline src/http-api/resources/files.js]: so maybe 'failedToImport' flag?

Comment (noffle) [inline src/http-api/resources/files.js]: @diasdavid: sorry, can you clarify what you're asking to be named to `failedToImport` here?

File: src/http-api/resources/files.js (hunk 2 above)
Comment (noffle) [inline src/http-api/resources/files.js]: I removed this -- seems like it was unnecessary!

File: src/http-api/resources/files.js (hunk 1 above)
Comment (nginnever) [inline src/http-api/resources/files.js]: maybe ```file``` could be called ```didParse``` to flag that the parser emitted a file event, ```written``` makes sense to me for the ```data``` event from the importer when data has been stored, or maybe ```imported``` is better since it is currently storing the number of files imported

Comment (noffle) [inline src/http-api/resources/files.js]: I took a stab at making this a bit clearer. Lemme know what y'all think!

**Comment by noffle** [comment]
Getting test timeouts -- don't mind the reverts (WIP)..

**Comment by diasdavid** [comment]
LGTM :) @noffle let me know if you have something to add (based on your last comment) or if I can merge it

File: test/cli-tests/test-bitswap.js [hunk 3]
```diff
@@ -12,7 +12,7 @@ const createTempNode = require('../utils/temp-node')
 const repoPath = require('./index').repoPath
 
 describe('bitswap', function () {
-  this.timeout(20000)
+  this.timeout(80000)
```
Comment (RichardLitt) [inline test/cli-tests/test-bitswap.js]: Could we make this an env variable? Seems weird to keep defining it everywhere.

**Comment by RichardLitt** [comment]
Might be good to add some failing tests, too - all of the ones here are positive.

File: test/cli-tests/test-bitswap.js (hunk 3 above)
Comment (dignifiedquire) [inline test/cli-tests/test-bitswap.js]: These are often different depending on the test, that's why they are set specifically in here.

File: src/http-api/resources/files.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    var serialize = ndjson.serialize()
+    // hapi doesn't permit object streams: http://hapijs.com/api#replyerr-result
+    serialize._readableState.objectMode = false
+
+    var fileAdder = request.server.app.ipfs.files.add()
+
+    fileAdder.on('data', (file) => {
+      serialize.write({
+        Name: file.path,
+        Hash: bs58.encode(file.multihash).toString()
```
Comment (dignifiedquire) [inline src/http-api/resources/files.js]: Please use `mh.toB58String(file.multihash)` from the `multihashes` package insteada

File: src/http-api/resources/files.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+const debug = require('debug')
+const log = debug('http-api:files')
+log.error = debug('http-api:files:error')
+
+exports = module.exports
+
+exports.add = {
+  handler: (request, reply) => {
+    if (!request.payload) {
+      return reply('Array, Buffer, or String is required').code(400).takeover()
```
Comment (dignifiedquire) [inline src/http-api/resources/files.js]: Are you sure `400` is the right status code for this?

File: test/http-api-tests/test-bitswap.js [hunk 6]
```diff
@@ -5,7 +5,7 @@ const expect = require('chai').expect
 
 module.exports = (httpAPI) => {
   describe('bitswap', function () {
-    this.timeout(20000)
+    this.timeout(80000)
```
Comment (dignifiedquire) [inline test/http-api-tests/test-bitswap.js]: why this unrelated change?

File: test/core-tests/test-swarm-node.js [hunk 7]
```diff
@@ -6,7 +6,7 @@ const expect = require('chai').expect
 const createTempNode = require('../utils/temp-node')
 
 describe('swarm', function () {
-  this.timeout(20000)
+  this.timeout(80000)
```
Comment (dignifiedquire) [inline test/core-tests/test-swarm-node.js]: why this unrelated change?

File: test/cli-tests/test-swarm.js [hunk 8]
```diff
@@ -9,7 +9,7 @@ const repoPath = require('./index').repoPath
 const _ = require('lodash')
 
 describe('swarm', function () {
-  this.timeout(20000)
+  this.timeout(80000)
```
Comment (dignifiedquire) [inline test/cli-tests/test-swarm.js]: why this unrelated change?

File: test/cli-tests/test-bitswap.js (hunk 3 above)
Comment (dignifiedquire) [inline test/cli-tests/test-bitswap.js]: why this unrelated change?

File: test/http-api-tests/test-files.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+          streamToPromise(form).then((payload) => {
+            api.inject({
+              method: 'POST',
+              url: '/api/v0/add',
+              headers: headers,
+              payload: payload
+            }, (res) => {
+              expect(res.statusCode).to.equal(200)
+              done()
```
Comment (dignifiedquire) [inline test/http-api-tests/test-files.js]: this should check the returned hash

File: test/http-api-tests/test-files.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
```
Comment (dignifiedquire) [inline test/http-api-tests/test-files.js]: this should check the returned hash

File: test/http-api-tests/test-swarm.js [hunk 11]
```diff
@@ -7,7 +7,7 @@ const createTempNode = require('../utils/temp-node')
 
 module.exports = (httpAPI) => {
   describe('swarm', function () {
-    this.timeout(20000)
+    this.timeout(80000)
```
Comment (dignifiedquire) [inline test/http-api-tests/test-swarm.js]: why this unrelated change?

File: test/cli-tests/test-bitswap.js (hunk 3 above)
Comment (noffle) [inline test/cli-tests/test-bitswap.js]: I could break this into another PR, but timeouts were making my local tests fail. :(

File: test/core-tests/test-swarm-node.js (hunk 7 above)
Comment (diasdavid) [inline test/core-tests/test-swarm-node.js]: ahah, saying it once is enough, I think :)

File: src/http-api/resources/files.js (hunk 4 above)
Comment (noffle) [inline src/http-api/resources/files.js]: Can do!

File: test/http-api-tests/test-files.js (hunk 9 above)
Comment (noffle) [inline test/http-api-tests/test-files.js]: Agreed!

File: test/http-api-tests/test-files.js (hunk 10 above)
Comment (noffle) [inline test/http-api-tests/test-files.js]: Done.

File: src/http-api/resources/files.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
```
Comment (RichardLitt) [inline src/http-api/resources/files.js]: Needs a period.

**Comment by noffle** [comment]
TODO: I still need to add those failing tests @RichardLitt mentioned

File: src/http-api/resources/files.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        Name: file.path,
+        Hash: multihash.toB58String(file.multihash)
+      })
+      written++
+    })
+
+    fileAdder.on('end', () => {
+      if (written === 0 && file) {
+        return reply({
+          Message: 'Failed to add files',
```
Comment (RichardLitt) [inline src/http-api/resources/files.js]: Need a period.

**Comment by noffle** [comment]
Assuming tests pass, should be good to go!

File: test/cli-tests/test-bitswap.js (hunk 3 above)
Comment (dignifiedquire) [inline test/cli-tests/test-bitswap.js]: please make it at least a separate commit

Comment (noffle) [inline test/cli-tests/test-bitswap.js]: I think they already are -- see 5d55b3cfe7d71f8d8b84f5926dbd54281fd38dc6. Did I miss one?

**Comment by noffle** [comment]
@diasdavid teamcity's failure is a cached one it looks like; it's just whining about a timeout. If you're okay with how this PR looks, then please feel free to merge!

**Comment by diasdavid** [comment]
@noffle looks good :) can we just get the commits squashed please?

**Comment by nginnever** [comment]
woo add command!

squashed commits and LGTM, gj @noffle

**Comment by nginnever** [comment]
@noffle one thing i forgot about this PR is that it doesn't address the cli looking for the daemon to be on.

I have written this code before and I just used it on your new add http-api resource and it works all ways!

You can plug this gist into this PR if you want, or I can open up a separate PR once this is merged.

https://gist.github.com/nginnever/39251f1d1572634e9699c875628dd00b

**Comment by noffle** [comment]
@nginnever That's safe to put in another PR, right? It sounds like you have a lot of context for what needs to happen for these changes -- if you could make that happen that'd be super awesome. :heart:

**Comment by diasdavid** [comment]
Master needs to be rebased onto this branch for the clean merge. @noffle could you do that?

@nginnever as @noffle suggested, please open a issue/PR with that missing part.

**Comment by diasdavid** [comment]
Only one more thing to go. This PR needs to update https://github.com/ipfs/js-ipfs/blob/master/src/cli/commands/files/add.js#L62, so that if the daemon is on, it doesn't fail. On that, it also needs cli tests to make sure add works when the daemon is on.

**Comment by nginnever** [comment]
can be closed in favor of <a href="https://github.com/ipfs/js-ipfs/pull/256">PR</a>

**Comment by diasdavid** [comment]
ok, thank you for merging them and documenting what is missing, I really like it :)
